kythe/cxx/common/kzip_writer_c_api.cc

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2018 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kythe/cxx/common/kzip_writer_c_api.h"
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "absl/status/status.h"
+#include "absl/strings/string_view.h"
+#include "kythe/cxx/common/index_writer.h"
+#include "kythe/cxx/common/kzip_writer.h"
+#include "kythe/cxx/common/status_or.h"
+#include "kythe/proto/analysis.pb.h"
+
+using kythe::IndexWriter;
+using kythe::KzipEncoding;
+using kythe::StatusOr;
+
+namespace {
+
+KzipEncoding ToKzipEncoding(int32_t encoding) {
+  KzipEncoding kzip_encoding = KzipEncoding::kJson;
+  switch (encoding) {
+    case KZIP_WRITER_ENCODING_JSON:
+      kzip_encoding = KzipEncoding::kJson;
+      break;
+    case KZIP_WRITER_ENCODING_PROTO:
+      kzip_encoding = KzipEncoding::kProto;
+      break;
+    default:
+      // fallthrough
+    case KZIP_WRITER_ENCODING_ALL:
+      kzip_encoding = kythe::KzipEncoding::kAll;
+      break;
+  }
+  return kzip_encoding;
+}
+
+int32_t CodeFromStatus(const absl::Status& s) {
+  const absl::StatusCode code = s.code();
+  return static_cast<int32_t>(code);
+}
+
+}  // namespace
+
+KzipWriter* KzipWriter_Create(const char* path, const size_t path_len,
+                              int32_t encoding, int32_t* create_status) {
+  *create_status = 0;
+  KzipEncoding kzip_encoding = ToKzipEncoding(encoding);
+  const absl::string_view path_view(path, path_len);
+
+  StatusOr<std::unique_ptr<IndexWriter>> writer_or =
+      kythe::KzipWriter::CreateUnique(path_view, kzip_encoding);
+  if (!writer_or.ok()) {
+    *create_status = CodeFromStatus(writer_or.status());
```

Comment:
I think dropping the message loses a lot of relevant information.  I'm not sure creating a full-fledged C-API status type is warranted, but at least logging the status may be?  Given that you'd want to do this consistently it may make sense to have CodeFromStatus (or a differently named equivalent) do so.

Reply:
Added logging.  We won't be swallowing the text message, though in the cases I've seen, the message isn't really very useful.